server/game/gameSystems/DeckSearchSystem.ts

```diff
@@ -0,0 +1,284 @@
+import Player from '../core/Player';
+import { shuffleArray } from '../core/utils/Helpers';
+
+type Derivable<T> = T | ((context: AbilityContext) => T);
+
+export interface IDeckSearchProperties extends IPlayerTargetSystemProperties {
+    targetMode?: TargetMode;
+    activePromptTitle?: string;
+    cardsToSearch?: number | ((context: AbilityContext) => number); // Number of cards to search through
+    selectCardCount?: number | ((context: AbilityContext) => number);
+    reveal?: boolean;
+    shuffle?: boolean | ((context: AbilityContext) => boolean);
+    title?: string;
+    immediateEffect?: GameSystem<IGameSystemProperties>; // What should be done with the selected cards?
+    message?: string;
+    uniqueNames?: boolean;
+    player?: Player;
+    choosingPlayer?: Player;
+    placeOnBottomInRandomOrder?: boolean;
+    messageArgs?: (context: AbilityContext, cards: Card[]) => any | any[];
+    selectedCardsHandler?: (context: AbilityContext, event: DeckSearchEvent, cards: Card[]) => void;
+    remainingCardsHandler?: (context: AbilityContext, event: DeckSearchEvent, cards: Card[]) => void;
+    cardCondition?: (card: Card, context: AbilityContext) => boolean;
+    chooseNothingImmediateEffect?: GameSystem<IGameSystemProperties>;
+}
+
+export class DeckSearchSystem extends PlayerTargetSystem<IDeckSearchProperties> {
+    public override readonly name = 'deckSearch';
+    public override readonly eventName = EventName.OnDeckSearch;
+    public override readonly costDescription: string = '';
+    public override readonly effectDescription: string = '';
+
+    protected override defaultProperties: IDeckSearchProperties = {
+        cardsToSearch: -1,
+        selectCardCount: 1,
+        targetMode: TargetMode.UpTo,
+        selectedCardsHandler: null,
+        remainingCardsHandler: null,
+        chooseNothingImmediateEffect: null,
+        shuffle: false,
+        reveal: true,
+        uniqueNames: false,
+        placeOnBottomInRandomOrder: true,
+        cardCondition: () => true
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(event): void { }
+
+    public override hasLegalTarget(context: AbilityContext, additionalProperties = {}): boolean {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
+        if (this.getAmount(properties.cardsToSearch, context) === 0) {
+            return false;
+        }
+        const player = properties.player || context.player;
+        return this.getDeck(player).length > 0 && super.canAffect(player, context);
+    }
+
+    public override generatePropertiesFromContext(context: AbilityContext, additionalProperties = {}): IDeckSearchProperties {
+        const properties = super.generatePropertiesFromContext(context, additionalProperties) as IDeckSearchProperties;
```

Comment:
Since cardCondition's default is a function that returns true, I'm not sure if we can check that super well. Does something like this work in TS?
if (Contract.assertTrue(properties.cardCondition !== (() => true))) {
            return;
        }

Reply:
Also, the method needs to return something unlike an event handler. Should I return the properties? Or should I maybe be handling the Contract checks elsewhere?

Reply:
Oh yeah fair, I guess if cardCondition doesn't default to null then we can't do much there. Nbd.

The method is supposed to return a properties object of whatever the `TProperties` for the class is - in this case, `IDeckSearchProperties`.